controlplane/kubeadm/internal/workload_cluster_etcd_test.go

```diff
@@ -0,0 +1,557 @@
+		},
+		Data: map[string]string{
+			clusterConfigurationKey: `
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+etcd:
+  local:
+    dataDir: /var/lib/etcd
+    imageRepository: "gcr.io/k8s/etcd"
+    imageTag: "0.10.9"
+`,
+		},
+	}
+
+	g := NewWithT(t)
+	scheme := runtime.NewScheme()
+	g.Expect(corev1.AddToScheme(scheme)).To(Succeed())
+
+	tests := []struct {
+		name                  string
+		objs                  []runtime.Object
+		imageRepo             string
+		imageTag              string
+		expectErr             bool
+		expectedClusterConfig string
+	}{
+		{
+			name:      "returns error if unable to find kubeadm-config",
+			expectErr: true,
+		},
+		{
+			name:      "updates the config map",
+			expectErr: false,
+			objs:      []runtime.Object{kubeadmConfig},
+			imageRepo: "gcr.io/imgRepo",
+			imageTag:  "v1.0.1-sometag.1",
+			expectedClusterConfig: `apiVersion: kubeadm.k8s.io/v1beta2
+etcd:
+  local:
+    dataDir: /var/lib/etcd
+    imageRepository: gcr.io/imgRepo
+    imageTag: v1.0.1-sometag.1
+kind: ClusterConfiguration
+`,
+		},
+		{
+			name:      "doesn't update the config map if there are no changes",
+			expectErr: false,
+			imageRepo: "gcr.io/k8s/etcd",
+			imageTag:  "0.10.9",
+			objs:      []runtime.Object{kubeadmConfig},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			fakeClient := fake.NewFakeClientWithScheme(scheme, tt.objs...)
+			w := &Workload{
+				Client: fakeClient,
+			}
+			ctx := context.TODO()
+			err := w.UpdateEtcdVersionInKubeadmConfigMap(ctx, tt.imageRepo, tt.imageTag)
+			if tt.expectErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).ToNot(HaveOccurred())
+			if tt.expectedClusterConfig != "" {
+				var actualConfig corev1.ConfigMap
+				g.Expect(w.Client.Get(
+					ctx,
+					ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem},
+					&actualConfig,
+				)).To(Succeed())
+				g.Expect(actualConfig.Data[clusterConfigurationKey]).To(Equal(tt.expectedClusterConfig))
+			}
+		})
+	}
+}
+
+func TestRemoveEtcdMemberFromMachine(t *testing.T) {
+	machine := &clusterv1.Machine{
+		Status: clusterv1.MachineStatus{
+			NodeRef: &corev1.ObjectReference{
+				Name: "cp1",
+			},
+		},
+	}
+	cp1 := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cp1",
+			Namespace: "cp1",
+			Labels: map[string]string{
+				labelNodeRoleMaster: "",
+			},
+		},
+	}
+	cp1DiffNS := cp1.DeepCopy()
+	cp1DiffNS.Namespace = "diff-ns"
+
+	cp2 := cp1.DeepCopy()
+	cp2.Name = "cp2"
+	cp2.Namespace = "cp2"
+
+	g := NewWithT(t)
+	scheme := runtime.NewScheme()
+	g.Expect(corev1.AddToScheme(scheme)).To(Succeed())
+	tests := []struct {
+		name                string
+		machine             *clusterv1.Machine
+		etcdClientGenerator etcdClientFor
+		objs                []runtime.Object
+		expectErr           bool
+	}{
+		{
+			name:      "does not panic if machine is nil",
+			expectErr: false,
```

Comment:
Yeah...I wasn't sure the best way to test this. I know gomega has the [Panic()](https://onsi.github.io/gomega/#panic) assertion. 
But if the `nil` checks are removed it's going to panic regardless because it's calling `removeMemberForNode(ctx, machine.Status.NodeRef.Name)`. And I decided that a panic is a failing test.
I added this test case just to document the behavior and not lose it.

Reply:
That is, if someone remove the nil checks, these tests gonna blow up.